src/main/java/jenkins/plugins/slack/user/EmailSlackUserIdResolver.java

```diff
@@ -0,0 +1,137 @@
+
+    public EmailSlackUserIdResolver(String authToken, CloseableHttpClient httpClient, List<MailAddressResolver> mailAddressResolvers) {
+        this.authToken = authToken;
+        this.httpClient = httpClient;
+        this.mailAddressResolvers = mailAddressResolvers != null ? mailAddressResolvers : MailAddressResolver.all();
+    }
+
+    public EmailSlackUserIdResolver(String authToken, CloseableHttpClient httpClient) {
+        this(authToken, httpClient, null);
+    }
+
+    @DataBoundConstructor
+    public EmailSlackUserIdResolver() {
+        this(null, null, null);
+    }
+
+    public String getAPIMethodURL() {
+        return LOOKUP_BY_EMAIL_METHOD_URL;
+    }
+
+    public void setAuthToken(String authToken) {
+        this.authToken = authToken;
+    }
+
+    public void setHttpClient(CloseableHttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    public void setMailAddressResolvers(List<MailAddressResolver> mailAddressResolvers) {
+        this.mailAddressResolvers = mailAddressResolvers;
+    }
+
+    protected String resolveUserId(User user) {
+        if (mailAddressResolvers == null) {
+            return null;
+        }
+
+       return mailAddressResolvers.stream()
+            .map(resolver -> resolver.findMailAddressFor(user))
+            .filter(StringUtils::isNotEmpty)
+            .map(this::resolveUserIdForEmailAddress)
+            .filter(StringUtils::isNotEmpty)
+            .findAny()
+            .orElse(null);
+    }
+
+    public String resolveUserIdForEmailAddress(String emailAddress) {
+        if (StringUtils.isEmpty(emailAddress) || StringUtils.isEmpty(authToken)) {
+            return null;
+        }
+
+        String slackUserId = null;
+        String url = String.format(LOOKUP_BY_EMAIL_METHOD_URL_FORMAT, authToken, emailAddress);
+        HttpGet getRequest = new HttpGet(url);
+        try (CloseableHttpResponse response = httpClient.execute(getRequest)) {
+            int responseCode = response.getStatusLine().getStatusCode();
+            if (HttpStatus.SC_OK == responseCode) {
+                HttpEntity entity = response.getEntity();
+                JSONObject slackResponse = new JSONObject(EntityUtils.toString(entity));
+                // additionally, make sure the JSON response contains an 'ok: true' entry
+                if (slackResponse.optBoolean(SLACK_OK_FIELD)) {
+                    JSONObject slackUser = slackResponse.getJSONObject(SLACK_USER_FIELD);
+                    slackUserId = slackUser.getString(SLACK_ID_FIELD);
+                }
+            }
+        } catch (IOException | ParseException | JSONException ex) {
+            LOGGER.log(Level.WARNING, "Error getting userId from Slack", ex);
+        }
+        return slackUserId;
+    }
+
+    @Extension
+    public static class DescriptorImpl extends SlackUserIdResolverDescriptor {
+
+        @Override
+        public String getDisplayName() {
+            return "Email User ID Resolver";
```

Comment:
should slack be in here?
```suggestion
            return "Slack email User ID Resolver";
```